### PR TITLE
fix!: deal statuses

### PR DESCRIPTION
**Pull request by elshan-eth**
- Add new SMALL_BALANCE status. It's when the balance is less than `targetWorkers * 2 epoches * price`.

### CODE REVIEW

File: src/deal/Deal.sol [hunk 1]
```diff
@@ -90,7 +90,7 @@ contract Deal is MulticallUpgradeable, WorkerManager, IDeal {
         uint256 prevEpoch = _globalCore().currentEpoch() - 1;
 
         DealStorage storage dealStorage = _getDealStorage();
-        dealStorage.maxPaidEpoch = prevEpoch;
+        dealStorage.maxPaidEpoch = 0;
```
Comment (folex) [inline src/deal/Deal.sol:93]: Let's leave a comment on why that makes sense?

File: src/deal/interfaces/IDeal.sol [hunk 2]
```diff
@@ -13,7 +13,8 @@ interface IDeal is IConfig, IWorkerManager {
         INSUFFICIENT_FUNDS,
         ACTIVE,
         ENDED,
-        NOT_ENOUGH_WORKERS
+        NOT_ENOUGH_WORKERS,
+        SMALL_BALANCE
```
Comment (folex) [inline src/deal/interfaces/IDeal.sol:17]: Let's comment what this means and how SMALL_BALANCE is different from INSUFFICIENT_FUNDS?

File: src/core/modules/market/Matcher.sol [hunk 3]
```diff
@@ -52,8 +52,7 @@ abstract contract Matcher is Offer, IMatcher {
         MatcherStorage storage matcherStorage = _getMatcherStorage();
 
         IDeal.Status dealStatus = deal.getStatus(); 
-        require(dealStatus != IDeal.Status.ENDED, "Matcher: deal is ended");
-        require(dealStatus != IDeal.Status.INSUFFICIENT_FUNDS, "Matcher: deal has insufficient funds");
+        require(dealStatus == IDeal.Status.ACTIVE || dealStatus == IDeal.Status.NOT_ENOUGH_WORKERS, "Matcher: deal is not active");
```
[the remark this answers is not in the document]
Comment (elshan-eth) [reply src/core/modules/market/Matcher.sol:55]: We need to match the deal if it is ACTIVE or NOT_ENOUGH_WORKERS. With other statuses, we don't need to match.

File: src/deal/Deal.sol [hunk 4]
```diff
@@ -164,16 +164,22 @@ contract Deal is MulticallUpgradeable, WorkerManager, IDeal {
...
+
+        if (maxPaidEpoch != 0 && _globalCore().currentEpoch() > dealStorage.maxPaidEpoch) {
             return Status.INSUFFICIENT_FUNDS;
+        } else if (freeBalance < _globalCore().minDealDepositedEpoches() * pricePerWorkerEpoch() * targetWorkers()) {
+            return Status.SMALL_BALANCE;
+        } else if (currentWorkerCount < minWorkers()) {
+            return Status.NOT_ENOUGH_WORKERS;
         } else {
             return Status.ACTIVE;
         }
```
Comment (folex) [inline src/deal/Deal.sol:179]: If status is INSUFFICIENT_FUNDS, you would calculate freeBalance and currentWorkerCount for nothing 🤔

Maybe do
```suggestion
        uint256 maxPaidEpoch = dealStorage.maxPaidEpoch;
        if (maxPaidEpoch != 0 && _globalCore().currentEpoch() > dealStorage.maxPaidEpoch) {
            return Status.INSUFFICIENT_FUNDS;
        } 
        
        uint256 freeBalance = getFreeBalance();
        if (freeBalance < _globalCore().minDealDepositedEpoches() * pricePerWorkerEpoch() * targetWorkers()) {
            return Status.SMALL_BALANCE;
        } 
        
        uint256 currentWorkerCount = getWorkerCount();
        if (currentWorkerCount < minWorkers()) {
            return Status.NOT_ENOUGH_WORKERS;
        }
       
        return Status.ACTIVE;
```

File: src/core/modules/market/Matcher.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (folex) [reply src/core/modules/market/Matcher.sol:55]: ah, I misread the condition. My bad.

**Review by folex — APPROVED** [review]

**Pull request merged**